src/Components/Server/src/Circuits/RevalidatingServerAuthenticationStateProvider.cs

```diff
@@ -0,0 +1,114 @@
+        private readonly TimeSpan _revalidationInterval;
+        private CancellationTokenSource _loopCancellationTokenSource = new CancellationTokenSource();
+
+        /// <summary>
+        /// Constructs an instance of <see cref="RevalidatingServerAuthenticationStateProvider"/>.
+        /// </summary>
+        /// <param name="loggerFactory">A logger factory.</param>
+        /// <param name="revalidationInterval">The interval between revalidation attempts.</param>
+        public RevalidatingServerAuthenticationStateProvider(
+            ILoggerFactory loggerFactory,
+            TimeSpan revalidationInterval)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            if (revalidationInterval == default)
+            {
+                throw new ArgumentException("The interval must be a nonzero value", nameof(revalidationInterval));
+            }
+
+            _logger = loggerFactory.CreateLogger<RevalidatingServerAuthenticationStateProvider>();
+            _revalidationInterval = revalidationInterval;
+
+            // Whenever we receive notification of a new authentication state, cancel any
+            // existing revalidation loop and start a new one
+            AuthenticationStateChanged += authenticationStateTask =>
+            {
+                _loopCancellationTokenSource?.Cancel();
+                _loopCancellationTokenSource = new CancellationTokenSource();
+                _ = RevalidationLoop(authenticationStateTask, _loopCancellationTokenSource.Token);
+            };
+        }
+
+        /// <summary>
+        /// Determines whether the authentication state is still valid.
+        /// </summary>
+        /// <param name="authenticationState">The current <see cref="AuthenticationState"/>.</param>
+        /// <returns>A <see cref="Task"/> that resolves as true if the <paramref name="authenticationState"/> is still valid, or false if it is not.</returns>
+        protected abstract Task<bool> ValidateAuthenticationStateAsync(AuthenticationState authenticationState);
+
+        private async Task RevalidationLoop(Task<AuthenticationState> authenticationStateTask, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var authenticationState = await authenticationStateTask;
+                if (authenticationState.User.Identity.IsAuthenticated)
+                {
+                    while (!cancellationToken.IsCancellationRequested)
+                    {
+                        try
+                        {
+                            await Task.Delay(_revalidationInterval, cancellationToken);
+                        }
+                        catch (TaskCanceledException)
+                        {
+                            return;
+                        }
+
+                        var isValid = await ValidateAuthenticationStateAsync(authenticationState);
```

Comment:
Does it make sense to propagate the cancellation token here? I realize that makes the error handling a little more challenging, but it's probably the right thing to do.

Reply:
Good point. Done.